### PR TITLE
Basic audio transitioning

### PR DESCRIPTION
**Pull request by daveystruijk**

### CODE REVIEW

File: contextproject-ms3/src/main/java/contextproject/audio/MixerProcessor.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      // Clipping
+      if (newValue > 1.0f) {
+        newValue = 1.0f;
+      } else if (newValue < -1.0f) {
+        newValue = -1.0f;
+      }
+      floatBuffer[i] = newValue;
+    }
+    return true;
+  }
```
Comment (daveystruijk) [inline contextproject-ms3/src/main/java/contextproject/audio/MixerProcessor.java]: Note: This entire class is currently unused, but is probably necessary later to combine both audio streams into one and do some limiting/compression, just like a real mixer does.

File: contextproject-ms3/src/main/java/contextproject/audio/MixerProcessor.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+  public float manipulateSignal(float signal) {
```
Comment (Roy-Graafmans) [inline contextproject-ms3/src/main/java/contextproject/audio/MixerProcessor.java]: Needs javadoc

File: contextproject-ms3/src/main/java/contextproject/audio/PlayerService.java [hunk 3]
```diff
@@ -22,57 +15,59 @@
...
-  private AudioDispatcher nextDispatcher;
-  private Thread currentThread;
-  private Thread nextThread;
   
   public Track currentTrack;
   public Track nextTrack;
+  
+  final int SAMPLE_RATE = 44100;
 
   protected PlayerService() throws EncoderException, LineUnavailableException {
```
Comment (Roy-Graafmans) [inline contextproject-ms3/src/main/java/contextproject/audio/PlayerService.java]: Needs javadoc

File: contextproject-ms3/src/main/java/contextproject/audio/PlayerService.java [hunk 4]
```diff
@@ -22,57 +15,59 @@
...
-    } catch (EncoderException e) {
-      e.printStackTrace();
-    } catch (LineUnavailableException e) {
+      currentProcessor.play(1.0);
+    } catch (EncoderException | LineUnavailableException e) {
       e.printStackTrace();
     }
   }
   
   public void transition() {
```
Comment (Roy-Graafmans) [inline contextproject-ms3/src/main/java/contextproject/audio/PlayerService.java]: Needs Javadoc

File: contextproject-ms3/src/main/java/contextproject/audio/PlayerService.java [hunk 5]
```diff
@@ -22,57 +15,59 @@
...
+        e.printStackTrace();
+      }
+      currentProcessor.setGain(1.0 - (i / 30.0));
+      nextProcessor.setGain((double) i / 30.0);
+    }
+    currentProcessor.unload();
+    currentProcessor = nextProcessor;
   }
   
   public void exit() {
```
Comment (Roy-Graafmans) [inline contextproject-ms3/src/main/java/contextproject/audio/PlayerService.java]: Needs Javadoc

File: contextproject-ms3/src/main/java/contextproject/audio/TrackProcessor.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  private WaveformSimilarityBasedOverlapAdd wsola;
+  private GainProcessor gainProcessor;
+  private AudioDispatcher dispatcher;
+
+  private double tempo;
+  private double currentTime;
+  private double pausedAt;
+  private double totalDuration;
+
+  public TrackProcessor(Attributes attributes) {
```
Comment (Roy-Graafmans) [inline contextproject-ms3/src/main/java/contextproject/audio/TrackProcessor.java]: Needs Javadoc

File: contextproject-ms3/src/main/java/contextproject/audio/TrackProcessor.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    try {
+      this.attributes = attributes;
+      this.format = Streamer.streamAudioFormat(attributes);
+    } catch (EncoderException e) {
+      e.printStackTrace();
+    }
+  }
+
+  public void load(Track track) {
```
Comment (Roy-Graafmans) [inline contextproject-ms3/src/main/java/contextproject/audio/TrackProcessor.java]: Needs Javadoc

File: contextproject-ms3/src/main/java/contextproject/audio/TrackProcessor.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+    this.track = track;
+    this.tempo = 1.0;
+    this.pausedAt = 0;
+    this.currentTime = 0;
+
+    setState(PlayerState.FILE_LOADED);
+  }
+
+  public void unload() {
```
Comment (Roy-Graafmans) [inline contextproject-ms3/src/main/java/contextproject/audio/TrackProcessor.java]: Needs Javadoc

File: contextproject-ms3/src/main/java/contextproject/audio/TrackProcessor.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    if(dispatcher != null) {
+      dispatcher.stop();
+    }
+    track = null;
+    setState(PlayerState.NO_FILE_LOADED);
+  }
+
+  public void play(double startGain) throws EncoderException, LineUnavailableException {
```
Comment (Roy-Graafmans) [inline contextproject-ms3/src/main/java/contextproject/audio/TrackProcessor.java]: Needs Javadoc

File: contextproject-ms3/src/main/java/contextproject/audio/TrackProcessor.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    dispatcher.addAudioProcessor(wsola);
+    dispatcher.addAudioProcessor(gainProcessor);
+    dispatcher.addAudioProcessor(audioPlayer);
+    
+    Thread thread = new Thread(dispatcher);
+    thread.start();
+    setState(PlayerState.PLAYING);
+  }
+
+  public void setGain(double gain) {
```
Comment (Roy-Graafmans) [inline contextproject-ms3/src/main/java/contextproject/audio/TrackProcessor.java]: Needs Javadoc

File: contextproject-ms3/src/main/java/contextproject/audio/TrackProcessor.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    gainProcessor.setGain(gain);
+  }
+  
+  private void setState(PlayerState state) {
+    this.state = state;
+  }
+
+  public static enum PlayerState {
```
Comment (Roy-Graafmans) [inline contextproject-ms3/src/main/java/contextproject/audio/TrackProcessor.java]: Needs Javadoc

File: contextproject-ms3/src/main/java/contextproject/audio/TrackProcessor.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    Thread thread = new Thread(dispatcher);
+    thread.start();
+    setState(PlayerState.PLAYING);
+  }
+
+  public void setGain(double gain) {
+    gainProcessor.setGain(gain);
+  }
+  
+  private void setState(PlayerState state) {
```
Comment (Roy-Graafmans) [inline contextproject-ms3/src/main/java/contextproject/audio/TrackProcessor.java]: Needs javadoc

**Pull request merged**